keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2164 @@
+are exhausted on that node, then the Pod cannot run elsewhere. The same applies
+to resources that are available on a certain subset of the nodes and those
+nodes are busy.
+
+Delayed allocation solves this by integrating allocation with Pod scheduling:
+an attempt to schedule a Pod triggers allocation of pending resources for nodes
+that the scheduler has deemed suitable. Scheduling the pod is then put on hold
+until all resources are allocated. This avoids scenarios where a Pod is
+permanently assigned to a node which can't fit the pod because of the pod's
+other resource requirements.
+
+### Sharing a single ResourceClaim
+
+Pods reference resource claims in a new PodSpec.ResourceClaims list. Each
+resource in that list can then be made available to one or more containers in
+that Pod. Depending on the capabilities defined in the
+ResourceClaim by the driver, a ResourceClaim can be used exclusively
+by one pod at a time or an unlimited number of pods. Support for additional
+constraints (maximum number of pods, maximum number of nodes) could be
+added once there are use cases for those.
+
+Users of a ResourceClaim don't need to be Pods. This KEP specifically supports
+Pods as users and describes how kube-scheduler and kubelet will deal with Pods
+that depend on a ResourceClaim, but the API and some custom resource driver
+might also be useful for controllers to manage resources without using those
+resources for Pods.
+
+### Ephemeral vs. persistent ResourceClaims lifecycle
+
+A persistent ResourceClaim has a lifecyle that is independent of any particular
+pod. It gets created and deleted by the user. This is useful for resources
+which are expensive to configure and that can be used multiple times by pods,
+either at the same time or one after the other. Such persistent ResourceClaims
+get referenced in the pod spec by name. When a PodTemplateSpec in an app
+controller spec references a ResourceClaim by name, all pods created by that
+controller also use that name and thus share the resources allocated for that
+ResourceClaim.
+
+But often, each Pod is meant to have exclusive access to its own ResourceClaim
+instance instead. To support such ephemeral resources without having to modify
+all controllers that create Pods, an entry in the new PodSpec.ResourceClaims
+list can also be a ResourceClaimTemplate. When a Pod gets created, such a
+template will be used to create a normal ResourceClaim with the Pod as owner
+with an
+[OwnerReference](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#OwnerReference)),
+and then the normal allocation of the resource takes place. Once the pod got
+deleted, the Kubernetes garbage collector will also delete the
+ResourceClaim.
+
+The difference between persistent and ephemeral resources for kube-scheduler
+and kubelet is that the name of the ResourceClaim needs to be determined
+differently: the name of an ephemeral ResourceClaim is `<pod>-<resource name in
```

Comment:
> assuming fixed names will end in tears.

It worked very well for generic ephemeral volumes :sweat_smile: 

The main advantage is that components and users are not forced to search for the ResourceClaim. This makes the code and running "kubectl" a lot easier.

Because the user controls his namespace, they can make sure that such name conflicts, while theoretically possible, do not occur in practice.

Reply:
FTR I hate this, but I won't fight too hard.

Do we fail to run the pod if there is a conflict?  We have to, right?  The claim can not be created, so the pod status can't be updated.

What is the name-truncation scheme when it is too long?

Reply:
> Do we fail to run the pod if there is a conflict? We have to, right? 

Yes. Usually, scheduling will already fail.

> What is the name-truncation scheme when it is too long?

There is no truncation scheme, such a situation is considered a user error. They need to be aware of the additional length constraint when choosing pod and claim name.

Reply:
yeeeuck!  

The user picks a perfectly valid pod name and a perfectly valid claim name.  They read the docs and followed the rules.  Then it fails during scheduling with what, an event?  

Oof, that makes me sad.

Reply:
But the comment for Template explicitly calls out the additional constraint. So if a user chooses to ignore that, they didn't follow the rules.

I need to check when exactly it gets checked. I think it is at Pod creation time, i.e. not during scheduling.

Reply:
It is during Pod creation time:
```
The Pod "my-csi-app-inline-volume-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" is invalid: 
* spec.volumes[0].name: Invalid value: "my-csi-volume-yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy": PVC name "my-csi-app-inline-volume-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-my-csi-volume-yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy": must be no more than 253 characters
```

I understand that this isn't ideal because it doesn't reject higher-level objects (Deployment, StatefulSet) as their pod names aren't known yet, but at least it doesn't waste too many cluster resources.